### PR TITLE
keep `path` & `repo` when `pinned`

### PR DESCRIPTION
**Pull request by t-bltg**
Fix https://github.com/JuliaLang/Pkg.jl/issues/3147.

- [x] add tests

### CODE REVIEW

File: test/pkg.jl [hunk 1]
```diff
@@ -926,4 +926,24 @@ end
     @test_throws Pkg.Types.PkgError("Package PackageSpec(\n  path = test_packages/Example\n  version = *\n) has neither name nor uuid") ensure_resolved(Pkg.Types.Context(), Pkg.Types.Manifest(), [p])
 end
 
+@testset "Issue #3147"
```
Comment (IanButterworth) [inline test/pkg.jl:929]:
```suggestion
@testset "Issue #3147" begin
```

Comment (t-bltg) [reply test/pkg.jl:929]: Yeah, sorry, draft for now ;)

Comment (IanButterworth) [reply test/pkg.jl]: Yep, just meaning to help the test run :)

File: src/Operations.jl [hunk 2]
```diff
@@ -1148,7 +1148,8 @@ function update_package_add(ctx::Context, pkg::PackageSpec, entry::PackageEntry,
             println(ctx.io, "`$(pkg.name)` is pinned at `v$(entry.version)`: maintaining pinned version")
         end
         return PackageSpec(; uuid=pkg.uuid, name=pkg.name, pinned=true,
-                           version=entry.version, tree_hash=entry.tree_hash)
+                           version=entry.version, tree_hash=entry.tree_hash,
```
Comment (fredrikekre) [inline src/Operations.jl:1151]: Is it really possible to pin everything here? What if I update the version in the Project.toml file in my local path? Perhaps if you pin a package that tracks a path it should only be the path that is pinned?

Comment (t-bltg) [reply src/Operations.jl:1151]: If you change the version in the `dev`ed path (**and pinned pkg**), then `Pkg.add("Example")` results in:
```
`Example` is pinned at `v0.5.3`: maintaining pinned version
   Resolving package versions...
    Updating `~/.julia/environments/v1.7/Project.toml`
  [7876af07] ~ Example v0.5.3 `[...]/Example.jl` ⚲ ⇒ v0.5.8 `[...]/Example.jl` ⚲
    Updating `~/.julia/environments/v1.7/Manifest.toml`
  [7876af07] ~ Example v0.5.3 `[...]/Example.jl` ⚲ ⇒ v0.5.8 `[...]/Example.jl` ⚲
```
I guess it is just consistent with `develop`, although the version is not `pinned` per se.

**Comment by t-bltg** [comment]
Bump, anyone.

**Comment by t-bltg** [comment]
@fredrikekre or @IanButterworth, could your review this PR or explain why it cannot be finalized ?

**Comment by t-bltg** [comment]
Hit this again, bump.

**Comment by t-bltg** [comment]
What can I do to get this PR reviewed / approved / merged ?
Thanks.

File: src/Pkg.jl [hunk 3]
```diff
@@ -296,7 +296,8 @@ const build = API.build
     Pkg.pin(pkgs::Union{PackageSpec, Vector{PackageSpec}}; io::IO=stderr, all_pkgs::Bool=false)
 
 Pin a package to the current version (or the one given in the `PackageSpec`) or to a certain
-git revision. A pinned package is never updated. To pin all dependencies set `all_pkgs=true`.
+git revision. A pinned package is never updated: if `pkg` is tracking a path, or a repository,
+those remain tracked.
```
Comment (IanButterworth) [inline src/Pkg.jl:300]: Did you mean to remove the info about pinning all?

Also the new sentence seems confusing. "Never updated" and "remain tracked" seem like they mean opposite things?

Comment (t-bltg) [reply src/Pkg.jl:300]: What is meant is that an update is not automatically updated. (If you look at the previous sentence "A pinned package is never updated." it was also confusing before).

**Comment by DilumAluthge** [comment]
@IanButterworth and @fredrikekre Can you review this?

**Review by IanButterworth — COMMENTED** [review]
Its not clear to me what the behavior of `up Foo` if Foo is pinned and tracking a repo here. The language around being pinned but still tracking is a little confusing. It might be good to explain that in the docstring a little more explicitly and add tests for `up` behavior?

**Comment by t-bltg** [comment]
@IanButterworth, I've clarified again and added more tests for `Pkg.update` behaviour under `Pkg.pin`.

Would you care to review again so that we can move forward on this PR ? Thanks.

File: src/Pkg.jl [hunk 4]
```diff
@@ -312,7 +312,9 @@ const build = API.build
     Pkg.pin(pkgs::Union{PackageSpec, Vector{PackageSpec}}; io::IO=stderr, all_pkgs::Bool=false)
 
 Pin a package to the current version (or the one given in the `PackageSpec`) or to a certain
-git revision. A pinned package is never updated. To pin all dependencies set `all_pkgs=true`.
+git revision. A pinned package is never automatically updated: if `pkg` is tracking a path,
+or a repository, those remain tracked in a frozen state). For example, if one wishes to update
+a pinned package in the current manifest, one must explicitly call `Pkg.pin` again.
```
Comment (IanButterworth) [inline src/Pkg.jl:317]: Does you have to call `Pkg.unpin` before `Pkg.update("Foo")`?

Sorry for the questions, I just don't use pin on tracked repos like this

Comment (t-bltg) [reply src/Pkg.jl:317]: Yes, in my experiments calling `Pkg.add` and `Pkg.update` on a pinned package is a noop.

One must either `Pkg.unpin` and update or one can also call `Pkg.pin` again, in order to update the manifest when `Project.toml` has changed (this is my use case: I track a path which is dev'ed, and pinned).

This is what I expect from pinning a package: freeze it until I explicitly want to work on it.

Comment (IanButterworth) [reply src/Pkg.jl:317]: Ok perhaps `one must unpin, update then pin the package again`

Comment (t-bltg) [reply src/Pkg.jl]: Thanks, (hopefully) done in https://github.com/JuliaLang/Pkg.jl/pull/3148/commits/414f9a0f40d098c13a7f1eb1ce70700feecddf2e.

File: src/Pkg.jl [hunk 5]
```diff
@@ -312,7 +312,9 @@ const build = API.build
     Pkg.pin(pkgs::Union{PackageSpec, Vector{PackageSpec}}; io::IO=stderr, all_pkgs::Bool=false)
 
 Pin a package to the current version (or the one given in the `PackageSpec`) or to a certain
-git revision. A pinned package is never updated. To pin all dependencies set `all_pkgs=true`.
+git revision. A pinned package is never automatically updated: if `pkg` is tracking a path,
+or a repository, those remain tracked in a frozen state). For example, if one wishes to update
+a pinned package in the current manifest, one must unpin, update then pin the package again.
```
Comment (IanButterworth) [inline src/Pkg.jl:317]:
```suggestion
or a repository, those remain tracked but will not update. 
To get updates from the origin path or remote repository the package must first be freed.
```

Final suggestion.. :)

**Review by IanButterworth — APPROVED** [review]

**Pull request merged**